app/services/topic_bookmarkable.rb

```diff
@@ -12,7 +12,16 @@ def self.serializer
   end
 
   def self.preload_associations
-    [:topic_users, :tags, { posts: :user }]
+    [:tags, { posts: :user }]
+  end
+
+  def self.perform_custom_preload!(topic_bookmarks, guardian)
```

Comment:
Instead of defining a custom preload method, I think the API we expose will be simpler if we support a `preload_scope` method which the preloader can then call.

https://github.com/rails/rails/blob/63d4f99e2bca562110d2597ba02541c400fd5b74/activerecord/lib/active_record/associations/preloader.rb#L96

Reply:
If I'm understanding correctly, this would require defining a scope on the Topic model, right? In this case, the record we need to preload varies based on the current_user. Essentially, we need to preload the result of this:

```
topic.topic_user.where(user_id: current_user)
```

I don't think we can express that as a scope or as a relation on the topic model 🤔

Reply:
O what I meant is doing something like this instead: 

```
bookmarks = Bookmark.where(bookmarkable_type: "Topic").to_a
ActiveRecord::Associations::Preloader.new(records: bookmarks, associations: :bookmarkable).call
ActiveRecord::Associations::Preloader.new(records: bookmarks.map(&:bookmarkable), associations: :topic_users, scope: TopicUser.where(user_id: 1)).call
```

I looked into adding a scope to `ActiveRecord::Associations::Preloader.new(records: bookmarks, associations: [bookmarkable: :topic_users])` but it seems like you can only provide a scope which will be applied to all the associations and nested associations being preloaded so the only way to make this work is to loop through the associations and provided a targeted scope for each which seems to be too complicated.

Reply:
I opened a topic on the Rails discussion forum to see whether a scope can be provided to target certain associations only: https://discuss.rubyonrails.org/t/providing-a-scope-for-nested-associations-when-preloading-with-activerecord-preloader/81245

Reply:
:+1: great, thanks @tgxworld. But I guess as far as this PR is concerned, we'll have to stick to the custom system for now?